### PR TITLE
[#164] First basic implementation of Pi estimation.

### PR DESCRIPTION
**Pull request by mminutoli**

### CODE REVIEW

File: examples/pi/pi.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+          double y = dist(G);
+          if ((x * x + y * y) < 1) {
+            ++counter;
+          }
+        }
+        return counter;
+      });
+
+  size_t count = shad::reduce(
+      shad::distributed_sequential_tag{}, counters.begin(), counters.end());
```
Comment (droccom) [inline examples/pi/pi.cc]:
```suggestion
      shad::distributed_parallel_tag{}, counters.begin(), counters.end());
```

File: examples/pi/pi.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+namespace shad {
+
+int main(int argc, char *argv[]) {
+
+  shad::array<size_t, 128> counters;
+
+  size_t numberOfPoints = 1e12;
+  size_t numberOfPointsPerSim = numberOfPoints / counters.size();
+
+  shad::generate(
```
Comment (droccom) [inline examples/pi/pi.cc]: I would prefer a `transform` over an array of seeds, so that the output does not change with the execution policy.

Comment (mminutoli) [reply examples/pi/pi.cc]: Do you want me to change it in a transform?

File: examples/pi/pi.cc [hunk 3]
```diff
@@ -37,17 +37,26 @@ namespace shad {
 int main(int argc, char *argv[]) {
 
   shad::array<size_t, 128> counters;
+  shad::array<size_t, 128> seeds;
 
-  size_t numberOfPoints = 1e12;
+  std::mt19937_64 seed_gen;
+  shad::generate(
+      shad::distributed_sequential_tag{},
```
Comment (droccom) [inline examples/pi/pi.cc]: Did you check that different sequences are generated on different localities? (this is not guaranteed by the semantics of `generate`).

**Comment by droccom** [comment]
I see numeric issues in some platforms, we should at least:
- replace `size_t` with `uint64_t` everywhere
- force the initial value for the `reduce` to be `uint64_t` (0 is sometimes inferred as a 32-bit type)

File: examples/pi/pi.cc [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      });
+
+  size_t numberOfPoints = 1e10;
+  size_t numberOfPointsPerSim = numberOfPoints / counters.size();
+
+  shad::transform(
+      shad::distributed_parallel_tag{},
+      seeds.begin(), seeds.end(),
+      counters.begin(),
+      [=](const uint64_t & S) -> size_t {
```
Comment (droccom) [inline examples/pi/pi.cc]: Better to use `uint64_t` instead of `size_t` for counters, I see overflows on my laptop (probably due to 32-bit `size_t`).

**Comment by droccom** [comment]
closes #164.

**Pull request merged**